enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,513 @@
+- Be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers).
+
+*Note: static resource creation is delegated to CVO (SA, RBAC, etc.)*
+
+#### CVO managemnet
+
+The operator image will be built and included in every release payload.
+It is expected to be deployed and running at Kubernetes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).
+
+Proposed `CCMO` manifests runlevel is `26`.
+
+#### Cloud provider management
+
+Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.
+
+The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, `ServiceMonitors` for metrics for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.
+
+Forked provider repositories will also contain cloud-provider specific code, and a set of binaries for each of them to build and run. Each provider will be built inside a single image, which will be included in the release payload. The responsibility for the operator will be to choose which provider image should be used in a cluster at any moment.
+
+#### CCM migration from KCM pod
+
+Currently the cloud-specific controller loops are running inside the `kcm` pod. It is ok to run in-tree and out-of-tree `CCM` simultaniously in a cluster for a short amount of time, or have a short amount of downtime between the moment in-tree plugin shuts down, and out-of-tree starts, but this gap should not be experienced by components, which rely on `CCM` functionality. Those are:
+- `opneshift-ingress` with `Route` resource management - non-operational without `CCM`.
+- `kubelet` - can't initialize a new `Node` without `CCM`.
+
+Although `CCM` controller loops itself does not require HA setup, the in-tree storage plugins are removed together with the `CCM`. This results in `CCM` dependency on the [cluster-storage-operator](https://github.com/openshift/cluster-storage-operator), and it's ability to support the provider CSI plugin and perform migration from in-tree storage plugin to the CSI one.
+
+#### Credentials management
+
+The operator is expected to be integrated with the [cloud-credentials-operator](https://github.com/openshift/cloud-credential-operator), and request fine grained credentials for the cloud components.
```

Comment:
This will need rewording but I think that was always our intention. CVO will create a number of `CredentialsRequests` and the CCO will provision appropriately. I think this is described below in `Post-install phase`